src/main/java/com/salesforce/cdp/queryservice/util/TokenHelper.java

```diff
@@ -276,4 +337,70 @@ private static Response un_pw_login(
+        byte[] grantTypeSegment = (Constants.GRANT_TYPE_NAME + Constants.TOKEN_ASSIGNMENT + grantType)
+                .getBytes(StandardCharsets.UTF_8);
+        byte[] jwsSegment = (Constants.ASSERTION + Constants.TOKEN_ASSIGNMENT +
+                createJwt(clientId, userName, privateKey, audience)).getBytes(StandardCharsets.UTF_8);
+        byte[] separator = Constants.TOKEN_SEPARATOR.getBytes(StandardCharsets.UTF_8);
+
+        byte[] body = Bytes.concat(
+                grantTypeSegment, separator, jwsSegment
+        );
+        try {
+            RequestBody requestBody = RequestBody.create(body, MediaType.parse(Constants.URL_ENCODED_CONTENT));
+            Map<String, String> headers = Collections.singletonMap(Constants.CONTENT_TYPE, Constants.URL_ENCODED_CONTENT);
+            Request request = HttpHelper.buildRequest(Constants.POST, tokenUrl, requestBody, headers);
+            Response response = client.newCall(request).execute();
+            if (!response.isSuccessful()) {
+                log.error("login with user credentials failed with status code {}", response.code());
+                HttpHelper.handleErrorResponse(response, Constants.ERROR_DESCRIPTION);
+            }
+            return response;
+        } catch (IOException e) {
+            log.error("login with user credentials failed", e);
+            throw new TokenException(FAILED_LOGIN, e);
+        } finally {
+            fillArray(grantTypeSegment, (byte)0);
+            fillArray(jwsSegment, (byte)0);
+            fillArray(separator, (byte)0);
+            fillArray(body, (byte)0);
+        }
+    }
+
+    private static String createJwt(String clientId, String userName, String privateKey, String audience) throws TokenException {
+        Instant now = Instant.now();
+        String jwtToken = null;
+        try {
+            RSAPrivateKey rsaPrivateKey = getPrivateKey(privateKey);
+            jwtToken = Jwts.builder()
+                    .setIssuer(clientId)
+                    .setSubject(userName)
+                    .setAudience(audience)
+                    .setIssuedAt(Date.from(now))
+                    .setExpiration(Date.from(now.plus(5l, ChronoUnit.MINUTES)))
+                    .signWith(rsaPrivateKey, SignatureAlgorithm.RS256)
+                    .compact();
+        } catch (Exception e) {
+            log.error("JWT assertion creation failed", e);
+            throw new TokenException(JWT_CREATION_FAILURE, e);
+        }
+
+        return jwtToken;
+    }
+
+    private static RSAPrivateKey getPrivateKey(String rsaPrivateKey) throws NoSuchAlgorithmException, InvalidKeySpecException {
+        rsaPrivateKey = rsaPrivateKey.replace("-----BEGIN PRIVATE KEY-----", "");
+        rsaPrivateKey = rsaPrivateKey.replace("-----END PRIVATE KEY-----", "");
+        rsaPrivateKey = rsaPrivateKey.replaceAll("\\s", "");
+
+        PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(Base64.getDecoder().decode(rsaPrivateKey));
```

Comment:
Is only PKCS8EncodedKeySpec type is supported by Salesforce? Can we confirm it once